File: Binary Search/Search a 2D Matrix/solution.ts
function searchMatrix(matrix: number[][], target: number): boolean {
    let min = 0
    let max = matrix.length - 1
    while(min < max) {
        let mid = min + Math.ceil((max - min + 1)/2)
        if(matrix[mid][0] === target) {
            return true
        } else if(matrix[mid][0] > target) {
            max = mid - 1
        } else {
            min = mid
        }
    }
    if(matrix[min][0] > target) return false
    let row = min
    min = 0
    max = matrix[0].length - 1
    while(min < max) {
        let mid = min + Math.floor((max - min) / 2)
        if(matrix[row][mid] === target) return true
        else if(matrix[row][mid] > target) {
            max = mid - 1
        } else {
            min = mid + 1
        }
    }
    if(matrix[row][min] === target) return true
    return false
};

File: Binary Search/Search a 2D Matrix/solution.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m, n = len(matrix), len(matrix[0])
        start = 0
        end = m - 1
        while start < end:
            mid = start + math.ceil((end - start + 1) / 2)
            if matrix[mid][0] == target:
                return True
            if matrix[mid][0] > target:
                end = mid - 1
            else:
                start = mid
                
        row = start
        if matrix[row][0] > target:
            return False
        
        start = 0
        end = n - 1
        while start < end:
            mid = start + (end - start) // 2
            if matrix[row][mid] < target:
                start = mid + 1
            else:
                end = mid
        return matrix[row][start] == target